File: geosrc/binreaders/binreader.py
import os
import struct
import glob
import time


def read_values(filename, indices):
    # indices are sorted and unique
    values = []
    with open(filename, 'rb') as f:
        for index in indices:
            f.seek(index*4L, os.SEEK_SET)
            b = f.read(4)
            v = struct.unpack("@i", b)[0]
            if v == -2147483648:
                v = None
            values.append(v)
    return values


def getindices(n):
    return [10000+(i*3) for i in range(n)]


def timefn(method):
    def timed(*args, **kw):
        ts = time.time()
        result = method(*args, **kw)
        te = time.time()
        print('%s%s %2.2f sec' % (method.__name__, str(list(args)), te-ts))
        return result
    return timed


@timefn
def allmarspec(outer, inner):
    paths = glob.glob(r'D:\temp\sbg_10m\*.sbg')
    indices = getindices(inner)
    r = []
    for i in range(outer):
        r = [read_values(os.path.join(r'D:\temp\sbg_10m', path), indices) for path in paths]
    return r

if __name__ == '__main__':
    allmarspec(1, 10)  # 0.01 sec
    #allmarspec(10,10000)  # <26 sec
    allmarspec(10000, 10)  # <26 sec

File: geosrc/binreaders/binreader.go
package main

import (
    "fmt"
    "os"
    "syscall"
    "io/ioutil"
    "path"
    "time"
)

func bytes2int(b []byte) int32 {
    v := int32(0)
    for  i := 0; i < 4; i++ {
        v = v | (int32(b[i]) << (uint(8*i)))
    }
    return v
}

func readValues(indices []int, filename string) []int32 {
    results := make([]int32, len(indices))
    b := make([]byte, 4)
    f,_ := os.Open(filename)
    fd := syscall.Handle(f.Fd())
    //value := int32(0)
    for i, cellIndex := range indices {
        syscall.Seek(fd, int64(cellIndex*4), os.SEEK_SET) // f.Seek(int64(cellIndex*4), os.SEEK_SET)
        syscall.Read(fd, b) //f.Read(b)
        value := bytes2int(b) // around 10-20% faster then binary.Read
        //package "encoding/binary"
        //binary.Read(f,binary.LittleEndian, &value)
        if value != -2147483648 {
            results[i] = value
        } else {
            results[i] = 99999
        }
    }
    return results
}

func getindices(n int) []int {
    indices := make([]int, n)
    for i := range indices {
        indices[i] = 10000+(i*3)
    }
    return indices
}

func allmarspec (outer int, inner int) []int32 {
    dir := path.Join("D:","temp","sbg_10m")
    files,_ := ioutil.ReadDir(dir)
    indices := getindices(inner)
    results := make([]int32, len(files))
    start := time.Now()
    for i := 0; i<outer;i++ {
        for fi,fileInfo := range files {
            results[fi] = readValues(indices, path.Join(dir, fileInfo.Name()))[0]
        }
    }
    fmt.Printf("allmarspec %d %d took %s\n" ,outer, inner,time.Now().Sub(start).String())
    return results
}

func main() {
    //fmt.Println(readValues([]int{0,1,2,13}, "D:\\temp\\sbg_10m\\bathy_10m.sbg"))
    
    //fmt.Println(allmarspec(100,100)) // 1s
    //fmt.Println(allmarspec(1000,100)) // 10s
    fmt.Println(allmarspec(10,10000)) // 8s
    //fmt.Println(allmarspec(1,100000)) // 8s
    //fmt.Println(allmarspec(10000,10)) // 25s
}